service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/PertinentInformationObservationValueMapper.java

```diff
@@ -0,0 +1,191 @@
+
+import com.google.common.collect.ImmutableMap;
+
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Component
+public class PertinentInformationObservationValueMapper {
+    private static final Map<Class<? extends IBaseElement>, Function<IBaseElement, String>> VALUE_TO_PERTINENT_INFORMATION_FUNCTIONS =
+        ImmutableMap.<Class<? extends IBaseElement>, Function<IBaseElement, String>>builder()
+            .put(CodeableConcept.class, value -> processCodeableConcept((CodeableConcept) value))
+            .put(BooleanType.class, value -> processBooleanType((BooleanType) value))
+            .put(Range.class, value -> processRange((Range) value))
+            .put(Ratio.class, value -> processRatio((Ratio) value))
+            .put(TimeType.class, value -> processTimeType((TimeType) value))
+            .put(DateTimeType.class, value -> processDateTimeType((DateTimeType) value))
+            .put(Period.class, value -> processPeriod((Period) value))
+            .build();
+    private static final String CODEABLE_CONCEPT_VALUE_TEMPLATE = "Code Value: %s %s ";
+    private static final String BOOLEAN_VALUE_TEMPLATE = "Boolean Value: %s ";
+    private static final String RANGE_VALUE_TEMPLATE = "Range Value: Low %s %s High %s %s ";
+    private static final String RATIO_VALUE_TEMPLATE = "Ratio Value: %s %s %s / %s %s %s ";
+    private static final String TIME_VALUE_TEMPLATE = "Time Value: %s ";
+    private static final String DATE_TIME_VALUE_TEMPLATE = "DateTime Value: %s ";
+    private static final String PERIOD_VALUE_TEMPLATE = "Period Value: Start %s End %s ";
+    private static final String RANGE_PREFIX = "Range: ";
+    private static final String TEXT_PREFIX = "Text: ";
+    private static final String LOW_PREFIX = "Low: ";
+    private static final String HIGH_PREFIX = "High: ";
+
+    public String mapObservationValueToPertinentInformation(Type value) {
+        if (!isPertinentInformation(value)) {
+            throw new IllegalArgumentException(
+                String.format("Observation value of '%s' type can not be converted to pertinent information", value.getClass()));
+        }
+
+        return VALUE_TO_PERTINENT_INFORMATION_FUNCTIONS.get(value.getClass())
+            .apply(value);
+    }
+
+    public boolean isPertinentInformation(Type value) {
+        return VALUE_TO_PERTINENT_INFORMATION_FUNCTIONS.containsKey(value.getClass());
+    }
+
+    public String mapReferenceRangeToPertinentInformation(
+            Observation.ObservationReferenceRangeComponent observationReferenceRangeComponent) {
+        String result = RANGE_PREFIX;
+        if (observationReferenceRangeComponent.hasText()) {
+            result += TEXT_PREFIX + observationReferenceRangeComponent.getText() + StringUtils.SPACE;
+
+            if (observationReferenceRangeComponent.hasLow()) {
+                SimpleQuantity low = observationReferenceRangeComponent.getLow();
```

Comment:
73-80 and 83-92 are very similar, they could be made as function, and you could also consider using builder pattern for that kind of logic, in this case PertinentInformationBuilder

Reply:
I was thinking about that builder pattern: https://refactoring.guru/design-patterns/builder